### PR TITLE
Fix list of base theme fonts when a theme variation is applied. 

### PR DESCRIPTION
**Pull request by matiasbenedetto**
## What?
- Improve the list of base theme fonts in the font library modal.
- This fixes the issue of opening a different font than the one selected when a theme variation with repeated font slugs is applied.

## Why?
Fixes part of: https://github.com/WordPress/gutenberg/issues/59818
Fixes part of: https://github.com/WordPress/gutenberg/issues/59574

## How?
Using as a base theme fonts the list of fonts composed by the de-duplicated list of: 
 - the theme fonts in global styles (database)
 - the `theme.json` file font families

## Testing Instructions
1. activate a theme with theme style variations (example TT4 and TT3)
2. Switch the style variation.
3. Tap on the font family names from the sidebar.
4. Check that the font listed is the correct one.

### Before
Notice that the font selected in the sidebar is not the one displayed in the modal.

https://github.com/WordPress/gutenberg/assets/1310626/fc67bfaf-a46d-4c30-885d-6106071f7bbc

### After

https://github.com/WordPress/gutenberg/assets/1310626/7145d796-b4ff-46dd-ac48-a9b9a9e4a17a

### CODE REVIEW

**Review by getdave — COMMENTED** [review]
I took this for a spin. What I saw was that despite the correct font name being shown when clicking on the that family in the sidar, the modal seemed to be showing a _serif_ font in the preview when the font was supposed to be _sans_-serif.

It's unusual because the CSS `font-family` has `Instrument Sans` but the font doesn't seem to be applied.

https://github.com/WordPress/gutenberg/assets/444434/5f2eecbe-1892-4f11-9e78-0daf74cd6150

**Comment by matiasbenedetto** [comment]
@getdave, the behavior you mentioned is unrelated to this PR or the Font Library. It happens because of a pre-existing problem described in this issue: https://github.com/WordPress/gutenberg/issues/59965. Basically, the font faces defined in variations are not loaded in the browser in 'real-time' but just after the browser refresh.

**Comment by getdave** [comment]
> @getdave, the behavior you mentioned is unrelated to this PR or the Font Library. It happens because of a pre-existing problem described in this issue: #59965. Basically, the font faces defined in variations are not loaded in the browser in 'real-time' but just after the browser refresh.

Ok thanks for clarifying. In which case the test case you provided works as described 👍

File: packages/edit-site/src/components/global-styles/font-library-modal/context.js
Comment (scruffian) [inline packages/edit-site/src/components/global-styles/font-library-modal/context.js:124]: I think it would be good to add a comment explaining the difference between `themeFonts` and `baseThemeFonts`.

**Review by scruffian — APPROVED** [review]
This is testing well for me and the code makes sense. It would be good to add a comment to explain things more clearly.

File: packages/edit-site/src/components/global-styles/font-library-modal/context.js
Comment (mikachan) [reply packages/edit-site/src/components/global-styles/font-library-modal/context.js:124]: I agree with adding a comment here, or maybe the variable names could be changed to be more specific about the font origins (theme.json, global styles etc).

**Review by mikachan — APPROVED** [review]
This works well for me based on the testing instructions 👍

File: packages/edit-site/src/components/global-styles/font-library-modal/context.js
Comment (matiasbenedetto) [reply packages/edit-site/src/components/global-styles/font-library-modal/context.js:124]: Added here: https://github.com/WordPress/gutenberg/pull/59959/commits/fb62517fe3927f05680b2e9e94959339368da55a

**Review by creativecoder — COMMENTED** [review]
Following the steps on the video from https://github.com/WordPress/gutenberg/issues/59818, I can see that this fixes the issue with clicking on a font in the sidebar list and then seeing different font in the modal (e.g. clicking on "Jost" and then seeing the settings for "Cardo").

Here's a before and after of the Font Library (trunk compared to this branch), after selecting the "Ember" theme variation for TT4

| **Before** | **After** |
| - | - |
| image | image |

---

One strange thing I'm seeing for the Ember variation is that System Serif is listed twice using this branch. Is that related, or a different issue?

**Comment by matiasbenedetto** [comment]
> One strange thing I'm seeing for the Ember variation is that System Serif is listed twice using this branch. Is that related, or a different issue?

Thanks for the detailed review.

I've seen that. I found that one variation was using `System-serif` instead of `system-serif`, so it's not detected as repeated. I don't see that difference anymore in the core repo, so I think it was already fixed in TT4 and won't be a problem in 6.5. I guess it isn't worth adding extra processing to the slugs.

File: packages/edit-site/src/components/global-styles/font-library-modal/context.js
Comment (creativecoder) [inline packages/edit-site/src/components/global-styles/font-library-modal/context.js]:
```suggestion
	 * These inconsistencies can happen when the active theme fonts in global styles aren't defined in theme.json file as when a theme style variation is applied.
```

**Review by creativecoder — APPROVED** [review]
This fixes the issue where the theme font list in the sidebar doesn't match the font list in the font library modal after selecting a theme variation or typography preset.

Tested with both WP trunk and WP 6.4 + Gutenberg plugin activated.

Thanks for fixing this! I'm not very familiar with this part of the code, but the comments certainly help explain it.

Note one small comment typo.

**Comment by creativecoder** [comment]
> One strange thing I'm seeing for the Ember variation is that System Serif is listed twice using this branch. Is that related, or a different issue?

Can confirm it's the theme.json font slug typo that causes this, and it was fixed here: https://github.com/WordPress/wordpress-develop/commit/b617b5c79aefc70cf12ec8f555e5aa73ba17ff45

**Comment by matiasbenedetto** [comment]
@getdave @youknowriad This PR seems important to be included in the release candidate. Could we add it?

**Comment by oandregal** [comment]
The fix works as expected for me, and I cannot foresee uncontrolled side-effects: it's contained to the font library modal, essentially.

Besides the obvious, I've tested some edge cases, such as 1) declaring a font family in the theme.json (theme.json sans-serif) that wasn't defined in any variation 2) adding a font family in a variation (variation sans-serif) that wasn't defined anywhere else. This is how it works:

https://github.com/WordPress/gutenberg/assets/583546/d9972382-0987-4ed7-8d46-3cb7d968e608

File: packages/edit-site/src/components/global-styles/font-library-modal/context.js
Comment (mcsf) [inline packages/edit-site/src/components/global-styles/font-library-modal/context.js:124]: @matiasbenedetto: Since the value is now a union with `themeFonts`, are these still `baseThemeFonts` — or does the name need changing?

**Comment by getdave** [comment]
This PR was triaged by Editor release leads and contributors in WP Slack as part of the final WordPress 6.5 process.

See https://wordpress.slack.com/archives/C02QB2JS7/p1711369960505499 (requires registration).

The criteria used was:

- Was this bug identified during the RC cycle for WP 6.5?
- Is this bug directly related to one of the blessed features in this release?
- Would this bug cause existing website’s to break/fail upon upgrade to WP 6.5?
- What is the approximate scope of the impact? Scale: (Minimal) 1 - 5 (Severe).
- Is the fix introducing any new features or changes?

**Conclusion**: we propose that this fix _is_ included in WP 6.5.

**Reason**: bug affects the (blessed) Font Library feature in a moderately-severe way and this PR resolves in a clear and traceable manner.

We now await a final decision from the full release team.

**Comment by bph** [comment]
I just cherry-picked this PR to the release/18.0 branch to get it included in the next release: aab2adce1e

**Comment by getdave** [comment]
Looking at adding test coverage in https://github.com/WordPress/gutenberg/pull/60250

**Comment by getdave** [comment]
I just cherry picked this into WP 6.5 RC 4 branch https://github.com/WordPress/gutenberg/pull/60241/commits/975ee76e93adcdb4a5bd8b9bf87a589afd8d34ca

**Pull request merged**
